### PR TITLE
Merge sharded DF jobs and performance improvements

### PR DESCRIPTION
**Pull request by shreyakhajanchi**
1. Merged sharded jobs into 1 using Sharding Context
2. Removed two reshuffle steps
3. Added numBuckets based on max workers to a reshuffle step 
4. Added retry setting in spanner config
5. Added watchdog to kill long running Spanner transactions

### CODE REVIEW

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/sources/DataStreamIO.java [hunk 1]
```diff
@@ -172,6 +174,12 @@ public DataStreamIO withDirectoryWatchDuration(Duration directoryWatchDuration)
     return this;
   }
 
+  /** Set applyReshuffle to false inorder to skip the reshuffle step on datastream records. */
```
Comment (manitgupta) [inline v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/sources/DataStreamIO.java:177]: Add more docs here and when this should be used and why is this being added.

File: v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/SpannerTransactionWriterDoFn.java [hunk 2]
```diff
@@ -102,6 +105,11 @@ class SpannerTransactionWriterDoFn extends DoFn<FailsafeElement<String, String>,
   /* The run mode, whether it is regular or retry. */
   private final Boolean isRegularRunMode;
 
+  private transient AtomicLong transactionAttemptCount;
+  private transient AtomicBoolean isInTransaction;
+  private transient AtomicBoolean keepWatchdogRunning;
+  private transient Thread watchdogThread;
```
Comment (manitgupta) [inline v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/SpannerTransactionWriterDoFn.java:111]: Add a detailed doc on what does the WatchDog and why is it needed.

File: v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/utils/WatchdogRunnable.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import com.google.common.util.concurrent.Uninterruptibles;
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class WatchdogRunnable implements Runnable, Serializable {
```
Comment (manitgupta) [inline v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/utils/WatchdogRunnable.java:26]: UT for testing WatchdogRunnable?

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/sources/DataStreamIO.java (hunk 1 above)
Comment (shreyakhajanchi) [reply v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/sources/DataStreamIO.java]: done

File: v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/SpannerTransactionWriterDoFn.java (hunk 2 above)
Comment (shreyakhajanchi) [reply v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/SpannerTransactionWriterDoFn.java:126]: done

File: v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/utils/WatchdogRunnable.java (hunk 3 above)
Comment (shreyakhajanchi) [reply v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/utils/WatchdogRunnable.java:33]: done

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/sources/DataStreamIO.java [hunk 4]
```diff
@@ -172,6 +174,23 @@ public DataStreamIO withDirectoryWatchDuration(Duration directoryWatchDuration)
     return this;
   }
 
+  /**
+   * Set applyReshuffle to false in order to skip the reshuffle step on Datastream records.
+   *
+   * <p>In the Datastream to Spanner template, one reshuffle step is already performed to balance
```
Comment (manitgupta) [inline v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/sources/DataStreamIO.java:180]: Nit: DatastreamIO is generic we shouldn't reference a specific use-case for it. Just add that this to provide an additional flexibility in terms of how DatastreamIO outputs the records.

File: v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/sources/DataStreamIO.java [hunk 5]
```diff
@@ -172,6 +174,23 @@ public DataStreamIO withDirectoryWatchDuration(Duration directoryWatchDuration)
...
+   * the load across workers. Load testing showed that removing this second reshuffle did not impact
+   * performance but significantly reduced pipeline costs.
+   *
+   * <p>Since the cost of a pipeline depends on the amount of data processed, reshuffling increases
+   * data processing due to extra grouping and shuffling across workers. By skipping this extra
+   * reshuffle, we reduce the total data processed, leading to cost savings while maintaining
+   * performance.
```
Comment (manitgupta) [inline v2/datastream-common/src/main/java/com/google/cloud/teleport/v2/datastream/sources/DataStreamIO.java:187]: Makes sense, but you should also talk about trade-off between distributing records across workers and reshuffling costs and overhead. Maybe link the documentation to reshuffling here.

**Review by manitgupta — DISMISSED** [review]
LGTM with some minor comments on documentation

**Review by manitgupta — APPROVED** [review]

**Pull request merged**